libraries/freertos_plus/aws/ota/include/aws_ota_agent.h

```diff
@@ -271,6 +271,151 @@ typedef struct
+ *
+ * @param[in] C File context of the job being aborted
+ */
+typedef OTA_Err_t (* pxOTAPALCloseFileCallback_t)( OTA_FileContext_t * const C );
+
+
+/**
+ * @brief OTA create file to store received data callback function typedef.
+ *
+ * The user may register a callback function when initializing the OTA Agent. This
+ * callback is used to override the behavior of how a new file is created.
+ *
+ * @param[in] C File context of the job being aborted
+ */
+typedef OTA_Err_t (* pxOTAPALCreateFileForRxCallback_t)( OTA_FileContext_t * const C );
+
+
+/**
+ * @brief OTA Get Platform Image State callback function typedef.
+ *
+ * The user may register a callback function when initializing the OTA Agent. This
+ * callback is used to override the behavior of returning the platform image state
+ *
+ * @param[in] ulServerFileID File ID of the image received
+ */
+typedef OTA_PAL_ImageState_t (* pxOTAPALGetPlatformImageStateCallback_t)( uint32_t ulServerFileID );
+
+/**
+ * @brief OTA Reset Device callback function typedef.
+ *
+ * The user may register a callback function when initializing the OTA Agent. This
+ * callback is used to override the behavior of what happens when the OTA agent resets the device.
+ *
+ * @param[in] ulServerFileID File ID of the image received
+ */
+typedef OTA_Err_t (* pxOTAPALResetDeviceCallback_t)( uint32_t ulServerFileID );
+
+/**
+ * @brief OTA Set Platform Image State callback function typedef.
+ *
+ * The user may register a callback function when initializing the OTA Agent. This
+ * callback is used to override the behavior of how a platform image state is stored
+ *
+ * @param[in] ulServerFileID File ID of the image received
+ * @param[in] eState Platform Image State to be state
+ */
+typedef OTA_Err_t (* pxOTAPALSetPlatformImageStateCallback_t)( uint32_t ulServerFileID, OTA_ImageState_t eState ); 
+
+/**
+ * @brief OTA Write Block callback function typedef.
+ *
+ * The user may register a callback function when initializing the OTA Agent. This
+ * callback is used to override the behavior of how a block is written to a file.
+ *
+ * @param[in] C File context of the job being aborted
+ * @param[in] iOffset Offset into the file to write the data
+ * @param[in] pacData Data to be written at the offset
+ * @param[in] iBlocksize Block size of the data to be written
+ */
+typedef int16_t (* pxOTAPALWriteBlockCallback_t)( OTA_FileContext_t * const C, uint32_t iOffset, uint8_t * const pacData, uint32_t iBlockSize );
+
+/**
+ * @brief Custom Job callback function typedef.
+ *
+ * The user may register a callback function when initializing the OTA Agent. This
+ * callback will be called when the OTA agent cannot parse a job document.
+ *
+ * @param[in] pcJSON Pointer to the json document received by the OTA agent
+ * @param[in] ulMsgLen Length of the json document received by the agent
+ */
+typedef OTA_JobParseErr_t (* pxOTACustomJobCallback_t)( const char * pcJSON, uint32_t ulMsgLen );
```

Comment:
Do we need to also pass the File Context for the callback to populate it ?

Reply:
Since it's a custom job, there is no afr_ota key. The callback just returns the entire json doc without any parsed details. What will we put in the file context if we send it back?

Reply:
Ok, I was curious as to how this callback will be used, as the line [2606](https://github.com/aws/amazon-freertos/pull/704/files#diff-6778a459de8ac6acebe32ec87c57b11dR2606),  immediately after the callback checks if the context has a valid job name and fails if the job name is null. Shouldn't the job name be populated from callback?